src/Chainweb/Chainweb/MinerResources.hs

```diff
@@ -68,27 +77,85 @@ data MiningCoordination logger cas = MiningCoordination
     , _coordState :: !(TVar MiningState)
     , _coordLimit :: !Int
-    , _coord503s :: IORef Int }
+    , _coord503s :: !(IORef Int)
+    , _coordMiners :: !(TVar CachedPayloads)
+      -- ^ "Blessed" miners who have payloads pre-cached for fast work request
+      -- turn-around.
+    }
 
 withMiningCoordination
     :: Logger logger
     => logger
+    -> ChainwebVersion
     -> Bool
+    -> [Miner]
     -> CutDb cas
     -> (Maybe (MiningCoordination logger cas) -> IO a)
     -> IO a
-withMiningCoordination logger enabled cutDb inner
+withMiningCoordination logger ver enabled miners cdb inner
     | not enabled = inner Nothing
     | otherwise = do
+        cut <- _cut cdb
         t <- newTVarIO mempty
+        m <- initialPayloads cut miners >>= newTVarIO
         c <- newIORef 0
-        fmap snd . concurrently (prune t c) $ inner . Just $ MiningCoordination
-            { _coordLogger = logger
-            , _coordCutDb = cutDb
-            , _coordState = t
-            , _coordLimit = 2500
-            , _coord503s = c }
+        fmap thd . runConcurrently $ (,,)
+            <$> Concurrently (prune t c)
+            <*> Concurrently (cachePayloads m cut)
+            <*> Concurrently (inner . Just $ MiningCoordination
+                { _coordLogger = logger
+                , _coordCutDb = cdb
+                , _coordState = t
+                , _coordLimit = 2500
+                , _coord503s = c
+                , _coordMiners = m })
   where
+    cids :: [ChainId]
+    cids = HS.toList $ chainIds ver
+
+    pact :: PactExecutionService
+    pact = _webPactExecutionService . _webBlockPayloadStorePact $ view cutDbPayloadStore cdb
+
+    initialPayloads :: Cut -> [Miner] -> IO CachedPayloads
+    initialPayloads cut ms =
+        CachedPayloads . M.fromList <$> traverse (\m -> (m,) <$> fromCut m pairs) ms
+      where
+        pairs = map (\cid -> (cid, cut ^?! ixg cid)) cids
+
+    fromCut
+      :: Miner
+      -> [(ChainId, BlockHeader)]
+      -> IO (M.Map ChainId (T2 PayloadWithOutputs BlockCreationTime))
+    fromCut m cut = M.fromList <$> traverse (\(cid, bh) -> (cid,) <$> getPayload m bh) cut
+
+    getPayload :: Miner -> BlockHeader -> IO (T2 PayloadWithOutputs BlockCreationTime)
+    getPayload m parent = do
+        creationTime <- BlockCreationTime <$> getCurrentTimeIntegral
+        payload <- _pactNewBlock pact m parent creationTime
+        pure $ T2 payload creationTime
+
+    cachePayloads :: TVar CachedPayloads -> Cut -> IO ()
+    cachePayloads tcp old = do
+        new <- atomically $ _cutStm cdb
+        let oldChains = HM.keysSet $ _cutMap old
+            newChains = HM.keysSet $ _cutMap new
+            difChains = HS.toList  $ HS.difference newChains oldChains
+            parents   = map (\cid -> new ^?! ixg cid) difChains
+        CachedPayloads cached <- readTVarIO tcp
+        newCached <- M.traverseWithKey (g parents) cached
+        atomically . writeTVar tcp $ CachedPayloads newCached
```

Comment:
No other thread *writes* to this `TVar`, so it's safe to have its read/write separated. There is still a risk that the "best" Cut updates while all this is happening, though.